aptos-move/move-examples/veiled_coin/sources/veiled_coin.move

```diff
@@ -0,0 +1,371 @@
+        // verify a range proof that 'new_bal' is in [0, 2^{64}). Since 'new_bal' = 'bal' - 'amount' this implies that
+        // 'bal' - 'amount' >= 0 and therefore that 'bal' >= 'amount'.
+        assert!(bulletproofs::verify_range_proof(&private_balance, range_proof, MAX_BITS_IN_VALUE, VEILED_COIN_DST), ERANGE_PROOF_VERIFICATION_FAILED);
+
+        coin_store.private_balance = pedersen::commitment_as_compressed_point(&private_balance);
+
+        // Returns the withdrawn veiled coin
+        VeiledCoin<CoinType> {
+            private_value
+        }
+    }
+
+    //
+    // Tests
+    //
+
+    const SOME_RANDOMNESS_1: vector<u8> = x"e7c7b42b75503bfc7b1932783786d227ebf88f79da752b68f6b865a9c179640c";
+
+    #[test(myself = @veiled_coin, source = @0x1, destination = @0x2)]
+    fun basic_viability_test(
+        myself: signer,
+        source: signer,
+        destination: signer
+    ) acquires VeiledCoinMinter, VeiledCoinStore {
+        // Initialzie the veiled coin module
+        init_module(&myself);
+
+        // Set up two accounts so we can register a new coin type on them
+        let source_addr = signer::address_of(&source);
+        account::create_account_for_test(source_addr);
+        let destination_addr = signer::address_of(&destination);
+        account::create_account_for_test(destination_addr);
+
+        // Create some 1,000 fake money inside 'source'
+        coin::create_fake_money(&source, &destination, 1000);
+
+        // Split 500 and 500 between source and destination
+        coin::transfer<coin::FakeMoney>(&source, destination_addr, 500);
+
+        // Mint 100 veiled coins at source (requires registering a veiled coin store at 'source')
+        register<coin::FakeMoney>(&source);
+        mint<coin::FakeMoney>(&source, 100);
+
+        // Transfer 50 of these veiled coins to destination
+        let val = new_scalar_from_u64(50);
+        let rand = new_scalar_from_bytes(SOME_RANDOMNESS_1);
+        let rand = std::option::extract(&mut rand);
+
+        let comm = pedersen::new_commitment_for_bulletproof(&val, &rand);
+
+        // This will be the balance left at the source, that we need to do a range proof for
+        let source_randomness = scalar_neg(&rand);
+        let source_new_val = new_scalar_from_u64(50);
+        let (range_proof, source_new_comm) = bulletproofs::prove_range(&source_new_val, &source_randomness, MAX_BITS_IN_VALUE, VEILED_COIN_DST);
+
+        // Execute the veiled transaction: no one will be able to tell 50 coins are being transferred.
+        register<coin::FakeMoney>(&destination);
+        transfer_privately_to<coin::FakeMoney>(&source, destination_addr, pedersen::commitment_clone(&comm), &range_proof);
+
+        // Sanity check veiled balances
+        assert!(verify_opened_balance<coin::FakeMoney>(source_addr, 50, &source_randomness), 1);
+        assert!(verify_opened_balance<coin::FakeMoney>(destination_addr, 50, &rand), 1);
```

Comment:
Do i understand correctly? The recipient only knows they received some veiled coins. To be sure about the actual amount, they need the `amount` and `randomness` from the sender, perhaps shared off-chain.

Reply:
Indeed. Figuring out how to send the amount and randomness to the sender is what we're solving next.